src/Microsoft.IdentityModel.Abstractions/EventDetails.cs

```diff
@@ -0,0 +1,225 @@
+            DateTime value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            DateTimePropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.DateTime;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Sets a property on the event details.
+        /// </summary>
+        /// <param name="key">Property key.</param>
+        /// <param name="value">Property value.</param>
+        /// <param name="dataClassification">Optional parameter which indicates the classification of the data.</param>
+        /// <exception cref="ArgumentNullException">'key' is null.</exception>
+        public virtual void SetProperty(
+            string key,
+            double value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            DoublePropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.Double;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Sets a property on the event details.
+        /// </summary>
+        /// <param name="key">Property key.</param>
+        /// <param name="value">Property value.</param>
+        /// <param name="dataClassification">Optional parameter which indicates the classification of the data.</param>
+        /// <exception cref="ArgumentNullException">'key' is null.</exception>
+        public virtual void SetProperty(
+            string key,
+            Guid value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            GuidPropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.Guid;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Updates the state of the property data classification.
+        /// </summary>
+        /// <param name="key">Key of the property.</param>
+        /// <param name="dataClassification">DataClassification for the property.</param>
+        internal void AddDataClassificationIfNecessary(string key, DataClassification dataClassification)
+        {
+            if (dataClassification != DataClassification.SystemMetadata)
+                DataClassificationMapping[key] = dataClassification;
+            else
+                DataClassificationMapping.Remove(key);
+        }
+
+        /// <summary>
+        /// Internal enum to help differentiate acceptable dimension types.
+        /// </summary>
+        internal enum DictionaryTypeEnum
```

Comment:
>DictionaryTypeEnum [](http://example.com/codeflow?start=22&length=18)

This limits the types of values we support. Is it possible that some users might have events associated with user defined types?

Reply:
Changed direction here. Plus added a unit test to show an example extention.